samples/username-header/Core_6/Shared/UsernameMutator.cs

```diff
@@ -1,15 +1,26 @@
-﻿using System.Threading;
+﻿using System.Security.Principal;
+using System.Threading;
 using System.Threading.Tasks;
 using NServiceBus.Logging;
 using NServiceBus.MessageMutator;
 
 #region Mutator
 
 public class UsernameMutator :
-    IMutateOutgoingTransportMessages
+    IMutateOutgoingTransportMessages,
+    IMutateIncomingTransportMessages
 {
     static ILog log = LogManager.GetLogger("Handler");
 
+    public Task MutateIncoming(MutateIncomingTransportMessageContext context)
+    {
+        log.Info("Reading headers to set Thread.CurrentPrincipal");
+        var identityName = context.Headers["UserName"];
```

Comment:
So the approach to use the pipeline is good, but not sure if a mutator works if you would like to use DI. Likely the resolver needs to be initialized in the first stage.